src/librustc_data_structures/modified_set.rs

```diff
@@ -0,0 +1,133 @@
+use std::marker::PhantomData;
+
+use rustc_index::vec::Idx;
+
+use ena::undo_log::{Rollback, UndoLogs};
+
+#[derive(Copy, Clone, Debug)]
+enum UndoInner {
+    Add,
+    Drain { index: usize, offset: usize },
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct Undo<I>(UndoInner, PhantomData<I>);
+
+/// Tracks which indices have been modified and allows watchers to registered and notified of these
```

Comment:
Is this type useful on its own, or is it mostly used as part of one of the unification tables? It'd be good to link over to the other types that it's meant to be used with so people can gain some broader context into what is going on.